File: tzio-core/scala/src/main/scala/com/kineolyan/tzio/v1/scala/slot/DataSlot.scala
package com.kineolyan.tzio.v1.scala.slot

abstract class DataSlot extends InputSlot with OutputSlot
case class FilledSlot(value: Int) extends DataSlot {
  override def canRead: Boolean = true
  override def read(): (Int, DataSlot) = (value, EmptySlot())

  override def canWrite: Boolean = false
  override def write(value: Int): DataSlot = throw new RuntimeException("Cannot write into a filled slot")
}
case class EmptySlot() extends DataSlot {
  override def canRead: Boolean = false
  override def read(): (Int, DataSlot) = throw new RuntimeException("Cannot read from a empty slot")

  override def canWrite: Boolean = true
  override def write(value: Int): DataSlot = FilledSlot(value)
}

File: tzio-core/jdk18/src/main/java/com/kineolyan/tzio/v1/java/slot/DataSlot.java
package com.kineolyan.tzio.v1.java.slot;

import com.kineolyan.tzio.v1.java.TransactionalElement;

/**
 * Input/Output slot storing a single value.
 */
public class DataSlot implements InputSlot, OutputSlot, TransactionalElement {

	/** Value in the slot */
	private int value = 0;
	/** Flag marking that a value is currently stored */
	private boolean hasValue = false;
	/** Flag marking that the value has been consumed */
	private boolean hasValueAfterStep = false;

	/**
	 * Gets the value of the slot.
	 * @return the value
	 */
	public int getValue() {
		return this.value;
	}

	@Override
	public boolean canRead() {
		return this.hasValue && this.hasValueAfterStep;
	}

	@Override
	public int read() {
		assert canRead(): "Cannot read from this slot";
		this.hasValueAfterStep = false;
		return this.value;
	}

	@Override
	public boolean canWrite() {
		return !this.hasValue && !this.hasValueAfterStep;
	}

	@Override
	public void write(final int value) {
		assert canWrite(): "Cannot write into this slot";
		this.value = value;
		this.hasValueAfterStep = true;
	}

	@Override
	public void onStepEnd() {
		this.hasValue = this.hasValueAfterStep;
	}
}
